### ISSUE TITLE
Remove duplicated lines from subtitles

### ISSUE THREAD

**Issue by YakivGluck** [ytdl-org/youtube-dl#30833]
## Checklist

- [x] I'm reporting a feature request
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've searched the bugtracker for similar feature requests including closed ones

## Description

Subtitles downloaded from YouTube usually have duplicated lines. Removing such lines is more or less easy with additional scripts. But wouldn't it be great to have this feature integrated in youtube-dl with a separate option to activate it?

( [an example of mentioned scripts](https://gist.github.com/davidcortesortuno/64723e4262889f592def55c1927db651) )

**Comment by dirkf** [comment]
This is an interesting idea. We would have to formulate the rules for doing this and the linked Gist doesn't give me confidence that such rules are well understood.

From my limited acquaintance with the VTT format there seem to be two straightforward cases:
* duplicate text: consecutive cues with identical text where the cue period is identical: delete the second (and any further such cues);
* duplicate period: consecutive cues with identical text where the start of the second cue is equal to the end of the first cue, or possibly within the period of the first cue: extend the period of the first cue to the end of the second cue and delete the second (and similarly for any further such cues).

The first case is actually a degenerate special case of the second.

Is this an issue with other subtitle formats?

**Comment by GNtrazios** [comment]
i am interested in solving this issue but i need your help. 
Could u please share the urls of the videos of which vtt file have duplicated lines?

**Comment by GNtrazios** [comment]
Ιs anyone available for a few questions about some scripts?

**Comment by sga-13** [comment]
Is calling an external program within the scope?

On linux (or any posix system)

```sed '/^[[:space:]]*$/d' "$1" | sed 's/^[ \t]*//;s/[ \t]*$//' | awk '!($1 " " $2 in arr){print; arr[$1 " " $2] = 1}' ```

I ran this on the generated srt file, first bit deletes trailing white spaces on the lines and deletes empty lines, second bit checks for duplicates with awk

I ran the following (it has some extra bits not related to this issue, mostly merging timestamps in the same line as the words, and deleting the end timestamps)
```sed '/^[[:space:]]*$/d' "$1" | sed 's/^[ \t]*//;s/[ \t]*$//' | awk '!($1 " " $2 in arr){print; arr[$1 " " $2] = 1}' | awk '/^[[:alnum:][:space:]]*[[:alpha:]]/ {if (NR > 1) {print prev, $0;} else {print $0;}; prev = "";} {prev = $0;}' | cut -d' ' -f1,4- | sed -z 's/,[0-9]*//g' >| "$1".cleaned```

and it took only 15 ms ± 0.5 ms (averaged over 20 runs) (for context, the original file was 110 KiB, after duplicate removal, it became 65 KiB, with my additions, it became 25.2 KiB), it became, so on the systems with sed and awk, this is almost negligible overhead

for a long term solution, we can look for a python implementation

**Comment by dirkf** [comment]
It should be easy enough to implement a solution in Python once we all agree on what the problem is. Unfortunately no-one was able to spend time on defining the problem beyond [my original comment](https://github.com/ytdl-org/youtube-dl/issues/30833#issuecomment-1093903030) and one [slightly generic response](https://github.com/ytdl-org/youtube-dl/issues/30833#issuecomment-1384441837) to clarification requests from @GNtravios.

Probably _sed_ is one of the least suitable languages for expressing a requirement specification. Do I understand correctly that the proposed script is dealing with the first case in my original comment, linked above, and also removing trailing spaces and blank lines?

**Comment by sga-13** [comment]
yes, removing trailing spaces is important, as the duplicate lines sometimes have some extra white spaces, and deleting empty lines improve the awk loop (lesser total number lines)

yes I do understand having sed ruins portability, and this is one of those cases where I can also contribute with python, but I commented mostly because I saw no activity on issue. And I coincidentally just wrote the second bit today, and this issue came up.

I am trying to implement a efficient summarisation of youtube videos, hopefully with sumi, and reducing these lines make the output much better

**Comment by dirkf** [comment]
It's not the portability, just that a _sed_ script is pretty opaque to anyone other than its author.

Given a list of real-world cases that need to be fixed, we can add some heuristics to fix similar cases. In addition to the [cases I previously identified](https://github.com/ytdl-org/youtube-dl/issues/30833#issuecomment-1093903030), it's now suggested that white-space needs to be adjusted.

Verbatim problem subtitle text extracts that can be used as test-cases are needed.

**Comment by blshkv** [comment]
https://github.com/anandiamy/Batch-Single-Clean-Youtube-Vtt
that software (also Python-based) seems working fine, feel free to implement similar.
